library/alloc/src/boxed.rs

```diff
@@ -1166,8 +1166,13 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
     #[unstable(feature = "allocator_api", issue = "32838")]
     #[rustc_const_unstable(feature = "const_box", issue = "92521")]
     #[inline]
-    pub const unsafe fn from_raw_in(raw: *mut T, alloc: A) -> Self {
-        Box(unsafe { Unique::new_unchecked(raw) }, alloc)
+    pub const unsafe fn from_raw_in(ptr: *mut T, alloc: A) -> Self {
+        ub_checks::assert_unsafe_precondition!(
+            check_language_ub,
+            "Box::from_raw_in requires that the pointer is properly aligned",
+            (ptr: *mut () = ptr as *mut (), align: usize = align_of::<T>()) => ptr.is_aligned_to(align)
+        );
+        Box(unsafe { Unique::new_unchecked(ptr) }, alloc)
```

Comment:
Just in terms of UX, it's better that _all_ preconditions of an `unsafe fn` be asserted in it directly, instead of falling through to callees.

Reply:
So this should also check for non-null-ness? What about Box::from_raw (which calls Box::from_raw_in)? Should it get its own checks as well?

Reply:
> So this should also check for non-null-ness?

Yes

> What about Box::from_raw (which calls Box::from_raw_in)? Should it get its own checks as well?

These checks sometimes impose small but widespread compile time overhead. What I'd like is for the diagnostic to be clear and actionable, and reporting the name of the function that the user called seems like a pretty basic first step. So yes I would _prefer_ for `Box::from_raw` to also have basically a copy+paste of this check, but we may have to balance that against compile-time overhead. I can kick off a perf run to measure once you and I think this PR is in good order and it passes CI.

Reply:
Sounds like a plan :D